### ISSUE TITLE
Territory Subdivisions

### ISSUE THREAD

**Issue by Schultzer** [kipcole9/cldr#21]
Hi, @kipcole9
I would like to contribute to CLDR.

I'm in need of Territory Subdivisions data so I can list all the cities from each country or list all countries.

how would you like to see it implemented.? 
If you got any advise it would be nice.

Thanks for this library ❤️

**Comment by kipcole9** [comment]
@schultzer, thanks for the support and encouragement.

CLDR has some data related to subdivisions as I think you know - but as best I understand it, it encapsulates only countries (territories) down to major subdivisions.  The content is in [the core zip file](http://www.unicode.org/Public/cldr/31.0.1 <http://www.unicode.org/Public/cldr/31.0.1>):

* supplemental/subdivisions.xml which describes the containment (not included in this repo) and
* subdivisions/<locale>.xml which has the subdivision localised names (not included in this repo)

And the locale specific data is very very sparse too - localised country names are in the territories.json (included for each locale in this repo) but there is almost no localised data for the next level administrative region (state/province/….)

I have typically used [geonames](http://download.geonames.org/export <http://download.geonames.org/export>) for geo data since it has data down to “administrative region 4” which sounds like what you are after?  Especially if localised names are not important for you.

Let me know what you’re after?  This topic is valuable to me too so happy to collaborate,

**Comment by Schultzer** [comment]
@kipcole9 I see, I have to think more about this, I looked into http://geonames.org but that's not quite what I'm looking for, what I'm looking for is to list all countries and their cities with a specific locale :)
So I thought that using CLDR Territory Subdivisions would solve that.

similar to how twitter-cldr has World Territories where you can do 
```:us.localize(:es).as_territory # estados unidos```

I hope this clarify my request. 😄

**Comment by kipcole9** [comment]
Down to the country level thats pretty easy to add to ex_cldr.  But there is no city information (or any regional info below state/province level) in CLDR that I’m aware of.

If thats enough for you then I would suggest implementing it in a separate package that depends on ex_cldr.  If you reach the conclusion that its something you want to collaborate on, I’ll update ex_cldr to make the data available via `Cldr.Config.get_locale/1` which is the current standard approach and you could, if you wish, implement `ex_cldr_territories` that provides the localisation.

**Comment by kipcole9** [comment]
@Schultzer ping!  Any additional thoughts on the above?

**Comment by Schultzer** [comment]
@kipcole9 I been reading through ex_cldr to get the gist of it.
I got some additional questions before I try to start an implementation.

* What is the reasons behind consolidating each JSON to one huge file
why not just conform to the repository?

* I never wrote a lexer before is there a performance benefit compared to just parsing the JSON through Poisons and work on the map?

My initial thought is that it would be easier to read the code and implement more functionality if the JSON wasn't consolidated but I might be wrong.

**Comment by kipcole9** [comment]
@Schultzer, very fair question.  The way the lib is structured (and more clearly in the upcoming 0.8) is that `ex_cldr`'s primary responsibility is data organisation and locale management.  Then the dependent libs like `ex_cldr_numbers`, `ex_cldr_dates_times`, `ex_cldr_units` and `ex_cldr_lists` implement relevant functionality for that part of the CLDR domain.

### Assumptions about locale packaging

There are three reasons why each locale is packaged into a single json file:

1. Locales are downloaded at compile time and are not packaged in the hex package.  The full package is too large for hex if they were all included and given nobody is likely to use all 516 locales to do so would be wasteful.  So a single json file per locale makes it easy to implement compile-time locale downloading and its just a single file.

2. I do quite a lot of data transformation between the CLDR json repository and the final consolidated locale file.  The intent is to both simplify access to the data from a client lib and also to pre-compile where appropriate some parts of the data formats. For example, formats that have substitutions in them I precompile into a format that is easier/faster to use through the `Cldr.Substitution` module.

3. I want to make updating to new versions of CLDR as painless as possible.  Therefore having a consistent single locale file means that I take responsibility for data management and do so only in `ex_cldr`.  All client libs should not have to have visibility into this process unless there is a breaking change in CLDR.

### Cldr client library implementation for Territories

As to the implementation of localised regional information - my proposal is that I do the work required in `ex_cldr` to serve the regional information and you implement the client package such as `ex_cldr_territory` that uses it.  For my part I'll need to parse the xml and format it into json etc etc.  Your part would be a lib that uses the "standard" internal api of `Cldr.Config.get_locale/1` that will return a map of all of the locale information. This is part of my 'contract' with client libs and perhaps it helps explain why I went with a consolidated locale file approach - you won't have to deal with this part.

Basically a `ex_cldr` client library goes like this (and this is a good prompt to document it):

### Compile time responsibilities of a client lib

1. Call `Cldr.known_locales/0` to get the locale names of the configured locales
2. For each locale, call `Cldr.Config.get_locale/1` to get the `Map.t` of the locale information (you don't need to do any parsing or lexing)
3. Using that returned data, say `Cldr.Config.get_locale("es").territories`, generate one function per locale that returns the appropriate content

Here's the extract of that from the simplest lib, `ex_cldr_units`:

```
  # Generate the functions that encapsulate the unit data from CDLR
  for locale <- Cldr.Config.known_locales() do
    locale_data =
      locale
      |> Cldr.Config.get_locale
      |> Map.get(:units)

    for style <- @unit_styles do
      units = Map.get(locale_data, style)

      def available_units(unquote(locale), unquote(style)) do
        unquote(Map.keys(units) |> Enum.sort)
      end

      defp pattern_for(unquote(locale), unquote(style), unit) do
        {:ok, Map.get(unquote(Macro.escape(units)), unit)}
      end
    end
  end
```

### Runtime responsibilities of a client lib

1. Provide a public api (I have been using `to_string(element, locale: locale_name)` sort of format as you can see but that might not be appropriate for territories.  Process relevant arguments and call the function you created at compile time.

### Expectations

It is assumed (but not required) that `Cldr.Config.get_locale/1` is called only at compile time since it is reading a file and calling `Poison.decode/1` on it and also doing some data munging.  At compile team, to keep reasonable performance, each locale is loaded and decoded only once and cached in ETS so there is only a marginal performer hit having different libs call `Cldr.Config.get_locale/1` at compile time.

### Future work

I'm not happy with the complexity of the locale consolidation process, nor with the locale loading process.  The first step has been to isolate so that a client lib only has to call `Cldr.Config.get_locale/1` and they are returned a nice Elixir map to work with.   Part of the complexity is the source data itself - the packaging and organisation of the json content doesn't always lend itself to easy access and I wanted to abstract the navigation of the repository data away from a client lib developer.

Does that help at least understand the architecture?

**Comment by Schultzer** [comment]
@kipcole9 Thanks for the explanation.
I'll take a look at the `ex_cldr_units` package you have for some inspiration.

**Comment by Schultzer** [comment]
Hi @kipcole9,
I've now made an initial implementation of the [`ex_cldr_territories`](https://github.com/Schultzer/cldr_territories)

It would be nice with some feedback and advise to the API.

**Comment by Schultzer** [comment]
@kipcole9  I made a branch [`ex_cldr`](https://github.com/Schultzer/cldr/tree/add-territories-data) that retrieve and consolidate the primary data needed for  [`ex_cldr_territories`](https://github.com/Schultzer/cldr_territories).

I tried to make a pull request but I get a weird error from GitHub that the page is taking to long to load.

Anyway, I'm not quite sure if my implantation in [`ex_cldr`](https://github.com/Schultzer/cldr/tree/add-territories-data) is the right way to go.

Maybe it's should only compile the data if it detects [`ex_cldr_territories`](https://github.com/Schultzer/cldr_territories).

I'll like to hear your thoughts. 😄

**Comment by kipcole9** [comment]
@schultzer I had the same issue trying to look at the branch you built.  In parallel I had also started work.  On master branch now there are two new functions:

* `Cldr.Config.territory_containment/0`
* `Cldr.Config.territory_info/0`

Next, and I think last, step for the `ex_cldr` part is to build in the localised territory names.  As you pointed out, this is quite a lot of data and perhaps consolidating this into a single locale file isn't a good strategy.

I'm going to think about this over night and then we can discuss.

**Comment by Schultzer** [comment]
I just want to share some thoughts I had lately.

As mentioned before the `JSON` repository of [`unicode-cldr`](https://github.com/unicode-cldr/) is very sparse with its data compared to the main `XML` repository,
with that in mind, it makes me question if its the right way to go with the `JSON` repository or if we would gain more with using the main `XML` repository.

So let's imagine that we go with the main `XML` repository, what kind of possible positive improvement will this bring os? 

As I see it with would bring features like this
```elixir
iex> Cldr.Territory.info(:GB)
%{territory: "United Kingdom", population: 64, gdp: 2,800,000, currency: %{name: "British Pound", code: :GBP}, days_in_week: 4, first_day_of_week: [:mon, :sun], first_day_of_weekend: :sat, last_day_of_weekend: :sun, meas_system: :UK, paper_size: :A4}
```

this will open the possibility to easily know what currency, measurement, etc. are used in the given territory and build system on that.

Let me know what you think about this and what direction you see fits [`ex_cldr`](https://github.com/kipcole9/cldr/)


[Detailed Territory Currency Information](https://www.unicode.org/cldr/charts/31/supplemental/detailed_territory_currency_information.html)
[Territory Information](https://www.unicode.org/cldr/charts/31/supplemental/territory_information.html)

**Comment by kipcole9** [comment]
Good thought - and indeed the earliest versions of ex_cldr were built using the XML repository.  I found myself in XML parsing hell and thats why I changed to the JSON versions - even though the data organisation is different. 

At this point I'm not feeling strongly compelled to go back to the XML content for three main reasons:

1. It would require a complete rewrite of the data parsing code (mostly the content in /mix) and I'm really not feeling motivated for that :-) 
2. I initially got very excited by how much data is in CLDR.  But after a while my feeling has been that the focus for ex_cldr should be on functions that use the data, not the data itself.  Of course that doesn't mean another project couldn't use the data for its own purposes and present it differently 
3. The XML data uses a somewhat complex inheritance scheme, the JSON version normalises the inheritance tree so I don't have to think about that part.
4. The XML data tend to mix localised data with non-localised data. The JSON repo tends to put the non-localised data in `cldr_core/supplemental`.  Using your example:

```
iex> Cldr.Territory.info(:GB)
%{territory: "United Kingdom", population: 64, gdp: 2,800,000, currency: %{name: "British Pound", code: :GBP}, days_in_week: 4, first_day_of_week: [:mon, :sun], first_day_of_weekend: :sat, last_day_of_weekend: :sun, meas_system: :UK, paper_size: :A4}
```
This mixes non-localised `population`, `currency_code`, `days_in_week`, ... with localised content `territory`, `currency name` and so on.  If the data was organised this way we'd potentially end up with a lot of duplication and increased memory footprint.

In the current organisation of the data (largely following the JSON repo, but not always) the content you used in your example is (or could be) available as:

* `Cldr.Calendar.week_data/0` returns the day of the week information.  Its currently incomplete in the implementation on my side, but its the non-localised data so only one copy for all locales

* `Cldr.Currency.for_territory/2` makes good sense - I'll add that.  Note that the currency data itself does a reasonable job of tracking what currency was in effect during a period of time.  So `Cldr.Currency.for_territory/2` would require an `as_at` parameter or similar to fully cover the data.

* Measurement system info is in `cldr_core/supplemental/measurementData.json`.  It wouldn't be too hard to add.

So summing up:

1. No doubt the XML content is more complete - but parsing it is a pain and having got this far with the JSON repo I'd prefer to keep going unless there is a really compelling case to revisit the data structuring code.  One day I will have to do that since its not particularly lovely code; that would be the time to reconsider.

2. The JSON content does a better job of separating the localised and non-localised data which, for ex_cldr, seems a better organisation to use

**Comment by Schultzer** [comment]
First of all my bad, I hadn't realized that most of the data is there.
So it would be possible to make those functionalities that I would like.

So, in the end, it actually comes down to how we represent the data.
as I think you know what I really want to be available to achieve is basically get the relative information from a territory for each locale and the non-localized data.

it's would really be fantastic to have `Cldr.Currency.for_terrritory` 
and also in the future `Cldr.TelephoneCode.for_territory`, `Cldr.PostalCode.for_territory`

**Comment by kipcole9** [comment]
If you take a look at ex_cldr master (also on hex as version 0.9.0) you'll find three functions that get you most (but not all) of the way.  Your feedback will help work out the strategy for the rest.  Some questions:

1. Unit and Measurement preference data (i.e. unitPreferenceData.json and measurementData.json) could be added - its a bit fiddly so doesn't naturally rank high on my priorities - but I can adjust if its important to you.  calendarPreferenceData.json could also be added.

2. The week data could be duplicated into the territory data, but I'd prefer that was a client lib responsibility to I can avoid duplicated data.  OK with you?

```elixir
iex> Cldr.Config.territory_info |> Map.get(:GB)
%{
  currency: %{GBP: %{from: ~D[1694-07-27]}},
  gdp: 2788000000000,
  language_population: %{
    "bn" => %{population_percent: 0.67},
    "cy" => %{official_status: "official_regional", population_percent: 0.77},
    "de" => %{population_percent: 6},
    "el" => %{population_percent: 0.34},
    "en" => %{official_status: "official", population_percent: 99},
    "fr" => %{population_percent: 19},
    "ga" => %{official_status: "official_regional", population_percent: 0.026},
    "gd" => %{
      official_status: "official_regional",
      population_percent: 0.099,
      writing_percent: 5
    },
    "it" => %{population_percent: 0.34},
    "ks" => %{population_percent: 0.19},
    "kw" => %{population_percent: 0.0031},
    "ml" => %{population_percent: 0.035},
    "pa" => %{population_percent: 0.79},
    "sco" => %{population_percent: 2.7, writing_percent: 5},
    "syl" => %{population_percent: 0.51},
    "yi" => %{population_percent: 0.049},
    "zh-Hant" => %{population_percent: 0.54}
  },
  literacy_percent: 99,
  population: 64430400,
  telephone_country_code: 44
}

iex> Cldr.Config.territory_containment
%{
  "053": [:AU, :NF, :NZ],
  "143": [:TM, :TJ, :KG, :KZ, :UZ],
  "155": [:AT, :BE, :CH, :DE, :FR, :LI, :LU, :MC, :NL],
  UN: [:AD, :AE, :AF, :AG, :AL, :AM, :AO, :AR, :AT, :AU, :AZ, :BA, :BB, :BD,
       :BE, :BF, :BG, :BH, :BI, :BJ, :BN, :BO, :BR, :BS, :BT, :BW, :BY, :BZ,
       :CA, :CD, :CF, :CG, :CH, :CI, :CL, :CM, :CN, :CO, :CR, :CU, :CV, :CY,
       :CZ, :DE, :DJ, :DK, ...],
  "005": [:AR, :BO, :BR, :CL, :CO, :EC, :FK, :GF, :GY, :PE, :PY, :SR, :UY, :VE],
  "029": [:AG, :AI, :AW, :BB, :BL, :BQ, :BS, :CU, :CW, :DM, :DO, :GD, :GP, :HT,
          :JM, :KN, :KY, :LC, :MF, :MQ, :MS, :PR, :SX, :TC, :TT, :VC, :VG, :VI],
  "151": [:BG, :BY, :CZ, :HU, :MD, :PL, :RO, :RU, :SK, :UA],
  "142": [:"145", :"143", :"030", :"034", :"035"],
  "021": [:BM, :CA, :GL, :PM, :US],
  "057": [:FM, :GU, :KI, :MH, :MP, :NR, :PW],
  "018": [:BW, :LS, :NA, :SZ, :ZA],
  "014": [:BI, :DJ, :ER, :ET, :KE, :KM, :MG, :MU, :MW, :MZ, :RE, :RW, :SC, :SO,
          :SS, :TZ, :UG, :YT, :ZM, :ZW],
  "145": [:AE, :AM, :AZ, :BH, :CY, :GE, :IL, :IQ, :JO, :KW, :LB, :OM, :PS, :QA,
          :SA, :SY, :TR, :YE],
  EZ: [:AT, :BE, :CY, :DE, :EE, :ES, :FI, :FR, :GR, :IE, :IT, :LT, :LU, :LV,
       :MT, :NL, :PT, :SI, :SK],
  "154": [:GG, :IM, :JE, :AX, :DK, :EE, :FI, :FO, :GB, :IE, :IS, :LT, :LV, :NO,
          :SE, :SJ],
  "419": [:"013", :"029", :"005"],
  "061": [:AS, :CK, :NU, :PF, :PN, :TK, :TO, :TV, :WF, :WS],
  "019": [:"021", :"013", :"029", :"005"],
  "002": [:"015", :"011", :"017", :"014", :"018"],
  "003": [:"021", :"013", :"029"],
  "009": [:"053", :"054", :"057", :"061", :QO],
  "030": [:CN, :HK, :JP, :KP, :KR, :MN, :MO, :TW],
  "017": [:AO, :CD, :CF, :CG, :CM, :GA, :GQ, :ST, :TD],
  "054": [:FJ, :NC, :PG, :SB, :VU],
  "001": [:"019", :"002", :"150", :"142", :"009"],
  "039": [:AD, :AL, :BA, :ES, :GI, :GR, :HR, :IT, :ME, :MK, :MT, :RS, :PT, :SI,
          :SM, :VA, :XK],
  EU: [:AT, :BE, :CY, :CZ, :DE, :DK, :EE, :ES, :FI, :FR, :GB, :GR, :HR, :HU,
       :IE, :IT, :LT, :LU, :LV, :MT, :NL, :PL, :PT, ...],
  "034": [:AF, :BD, :BT, :IN, :IR, :LK, :MV, :NP, :PK],
  "035": [:BN, :ID, :KH, :LA, :MM, :MY, :PH, :SG, :TH, :TL, :VN],
  "015": [:DZ, :EG, :EH, :LY, :MA, :SD, :TN, :EA, :IC],
  "150": [:"154", :"155", :"151", :"039"],
  "013": [:BZ, :CR, :GT, :HN, :MX, :NI, :PA, :SV],
  "011": [:BF, :BJ, :CI, :CV, :GH, :GM, :GN, :GW, :LR, :ML, :MR, :NE, :NG, :SH,
          :SL, :SN, :TG],
  QO: [:AQ, :BV, :CC, :CX, :GS, :HM, :IO, :TF, :UM, :AC, :CP, :DG, :TA]
}

iex> Cldr.Config.week_info
%{
  first_day: %{
    LK: "mon",
    IR: "sat",
    LT: "mon",
    PT: "mon",
    AZ: "mon",
    SK: "mon",
    UY: "mon",
    SA: "sun",
    ES: "mon",
    UM: "sun",
    KG: "mon",
    BA: "mon",
    TH: "sun",
    MV: "fri",
    KE: "sun",
    FI: "mon",
    BE: "mon",
    BN: "mon",
    LB: "mon",
    GU: "sun",
    HK: "sun",
    MT: "sun",
    VI: "sun",
    MD: "mon",
    YE: "sun",
    MM: "sun",
    OM: "sat",
    EE: "mon",
    IS: "mon",
    HR: "mon",
    GE: "mon",
    BH: "sat",
    AX: "mon",
    BZ: "sun",
    NL: "mon",
    UA: "mon",
    PK: "sun",
    FJ: "mon",
    BS: "sun",
    CR: "mon",
    IT: "mon",
    IE: "sun",
    QA: "sat",
    GR: "mon",
    LA: "sun",
    RS: "mon",
    SY: "sat",
    SM: "mon",
    AS: "sun",
    ...
  },
  min_days: %{
    GG: 4,
    GI: 4,
    LT: 4,
    PT: 4,
    SK: 4,
    ES: 4,
    UM: 1,
    FI: 4,
    BE: 4,
    GU: 1,
    VI: 1,
    EE: 4,
    IS: 4,
    AX: 4,
    NL: 4,
    FJ: 4,
    IT: 4,
    IE: 4,
    GR: 4,
    SM: 4,
    GF: 4,
    JE: 4,
    IM: 4,
    DE: 4,
    VA: 4,
    RE: 4,
    MQ: 4,
    DK: 4,
    FO: 4,
    AN: 4,
    FR: 4,
    BG: 4,
    AD: 4,
    SE: 4,
    NO: 4,
    HU: 4,
    PL: 4,
    LU: 4,
    GB: 4,
    "001": 1,
    CZ: 4,
    LI: 4,
    MC: 4,
    CH: 4,
    GP: 4,
    AT: 4,
    SJ: 4,
    US: 1
  },
  weekend_end: %{
    "001": "sun",
    AE: "sat",
    AF: "fri",
    BH: "sat",
    DZ: "sat",
    EG: "sat",
    IL: "sat",
    IQ: "sat",
    IR: "fri",
    JO: "sat",
    KW: "sat",
    LY: "sat",
    MA: "sat",
    OM: "sat",
    QA: "sat",
    SA: "sat",
    SD: "sat",
    SY: "sat",
    TN: "sat",
    YE: "sat"
  },
  weekend_start: %{
    "001": "sat",
    AE: "fri",
    AF: "thu",
    BH: "fri",
    DZ: "fri",
    EG: "fri",
    IL: "fri",
    IN: "sun",
    IQ: "fri",
    IR: "fri",
    JO: "fri",
    KW: "fri",
    LY: "fri",
    MA: "fri",
    OM: "fri",
    QA: "fri",
    SA: "fri",
    SD: "fri",
    SY: "fri",
    TN: "fri",
    YE: "fri"
  }
}
```

**Comment by Schultzer** [comment]
First of all, it's beautiful ❤️ 

1. This is not top priority right now but I definitely reckon that it should be available at some point.
2. I Agree on no duplication.

one way could be that `ex_cldr_territories` handles the representation of the unit and measurement data, but I'm not sure if that would be the right way to go.

**Comment by kipcole9** [comment]
Cool, glad its on the right track.  Turns out that the postal code regexes were removed in CLDR 26 because they weren't being maintained so unlikely that they can be added.

I've added measurement system data now in `master` and will add to hex soon. I'll keep plugging away on:

* Calendar Preferences
* Unit Preference Data

and work out how to best handle the localised territory names - probably this weekend.

```elixir
iex> Cldr.Config.territory_info |> Map.get(:GB)
%{
  currency: %{GBP: %{from: ~D[1694-07-27]}},
  gdp: 2788000000000,
  language_population: %{
    "bn" => %{population_percent: 0.67},
    "cy" => %{official_status: "official_regional", population_percent: 0.77},
    "de" => %{population_percent: 6},
    "el" => %{population_percent: 0.34},
    "en" => %{official_status: "official", population_percent: 99},
    "fr" => %{population_percent: 19},
    "ga" => %{official_status: "official_regional", population_percent: 0.026},
    "gd" => %{
      official_status: "official_regional",
      population_percent: 0.099,
      writing_percent: 5
    },
    "it" => %{population_percent: 0.34},
    "ks" => %{population_percent: 0.19},
    "kw" => %{population_percent: 0.0031},
    "ml" => %{population_percent: 0.035},
    "pa" => %{population_percent: 0.79},
    "sco" => %{population_percent: 2.7, writing_percent: 5},
    "syl" => %{population_percent: 0.51},
    "yi" => %{population_percent: 0.049},
    "zh-Hant" => %{population_percent: 0.54}
  },
  literacy_percent: 99,
  measurement_system: "metric",
  paper_size: "A4",
  population: 64430400,
  telephone_country_code: 44,
  temperature_measurement: "metric"
}
```
